Initial GoogleSaleBanner implementation and domain type definition

daledupreez opened the pull request:
#### Changes proposed in this Pull Request

* This PR implements a new `GoogleSaleBanner` component that is intended for display in the domains listing page, but the banner should only be displayed when a number of conditions are all met:
  - The user's device is larger than a standard mobile device
  - We need to be running a sale for Google Workspace via a sale coupon
  - The user needs to be in a country that is supported by Google
  - The user needs to have at least one domain that doesn't have paid email service _and_ that domain supports  Google Workspace purchases (the most common requirement here is that the domain must use our nameservers)
* If _all_ of the above conditions are met and we have loaded the domain list, we show the banner on the domains listing page
* When the banner is displayed, we log a `calypso_email_google_workspace_sale_banner_impression` event to Tracks with a `cta_name` prop of `get-google-workspace`. (This implies we should not log this event for mobile devices.)
* When the user clicks on the banner, we log a `calypso_email_google_workspace_sale_banner_cta_click` event to Tracks.
* This PR also implements some new TS types for the domains package, with the most interesting being the `ResponseDomain` type, which I mostly defined using the code from `client/state/sites/domains/assembler.js`, which intentionally casts most of the values to specific types.
  - This piece can be split out into a separate PR if need be, but I think it will be well worth having these types available for documenting domain-related functions using the [JSDoc import syntax for types](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#import-types) -- we don't need to do this immediately, but we can start making updates as we touch files in various places
 * Note that I've implemented the banner in both the existing domain listing page and the new UI that's currently protected by the `domains/management-list-redesign` feature flag.
 * This PR also addresses what looks like a bug in the underlying `Banner` component for situations where we're on mobile and the user has specified `disableCircle` and `iconPath` - the icon component is duplicated. I'll tackle that underlying bug in a follow-up PR.

@Automattic/nomado, @poligilad-auto, and @saygunnyc, please share any feedback that you have on the banner.

##### Screenshots

_Note: the following screenshots were from testing with different discounts -- the discount is specified by the coupon configuration._

###### Desktop

<img width="1079" alt="Screenshot 2021-11-03 at 21 15 50" src="https://user-images.githubusercontent.com/3376401/140181312-370b78d1-82d8-4e31-b9d8-b01396e6f364.png">

###### Tablet

<img width="768" alt="Screenshot 2021-11-04 at 14 15 51" src="https://user-images.githubusercontent.com/3376401/140313368-032d4bcb-5137-40ce-9f71-515d757106d3.png">

###### Mobile

_As noted [here](https://github.com/Automattic/wp-calypso/pull/57530#issuecomment-960803138), we're not going to show the banner on mobile.

###### New design

<img width="1098" alt="Screenshot 2021-11-03 at 21 44 00" src="https://user-images.githubusercontent.com/3376401/140181808-f5be6e0d-19f1-4186-9550-b28e4534f99e.png">

#### Testing instructions

* Getting set up to test this requires access to a back end sandbox, as we need to tackle a few things:
  - You need to create a sales coupon for the test Store environment - see D69393-code for instructions
  - You need to enable the test Store environment
  - You need to modify one of the back end APIs to mock a test domain having WordPress.com nameservers - I have a sample in D69406-code
* Once you have all of those set up, load the UI via a local branch or the live branch
* Go to Upgrades -> Domains for a site that has at least one eligible domain - verify that you see the banner displayed with an eligible domain.
* Verify that the `calypso_email_google_workspace_sale_banner_impression` Tracks event is logged
* Load the domains listing page with `?flags=domains/management-list-redesign` and `?flags=-domains/management-list-redesign` set in the URL - this will load the new and old components, respectively.
* Click on "Get Google Workspace".
*  Verify that a `calypso_email_google_workspace_sale_banner_cta_click` event is logged to Tracks, and  you see the email comparison page, and that the `source` URL parameter has a value of `google-sale`.
  - For now, this page doesn't promote/auto-expand Google Workspace during sales - we will tackle that in a follow-up PR
  - Also note that #57516 addresses the details we display when Google Workspace is on sale
* Try the steps above for a site that doesn't have any domains eligible for the purchase - verify that you don't see the banner.
* Switch your back end APIs to the production API and ensure that you've disabled the Store sandbox/test mode.
* Work through the flow again and verify that you don't see the banner for the sites with and without eligible domains.

Related to #57516

daledupreez committed:
Initial GoogleSaleBanner implementation and domain type definition

daledupreez committed:
Add country check; re-add commented out check; clean up declaration for function

a8ci18n commented on client/my-sites/email/google-sale-banner/index.tsx:
```diff
@@ -0,0 +1,109 @@
+
+	return (
+		<Banner
+			callToAction={ translate( 'Get Google Workspace' ) }
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **20** times:
`translate( 'Google Workspace' )` *ES Score: 8*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/try/google-sale-banner)

saygunnyc commented:
> @Automattic/nomado, @poligilad-auto, and @saygunnyc, please share any feedback that you have on the banner.

I want to defer to @poligilad-auto as this is her territory. 😄

poligilad-auto commented:
Thanks for the ping. Seems fine to me in general @daledupreez.

A few comments:
1. This is more related to the actual banner component, I noticed that the CTA is misaligned, it's too close to the right. Check this out. I don't know if this is something you can address.
![image](https://user-images.githubusercontent.com/83961704/139834075-ab397388-cbd2-46f0-b4a8-8b14feab662c.png)
2. What happens for users that have more than 1 eligible domain (I know this is an edge case 🙂)? Do we choose one of those domains to display the banner for?

daledupreez commented:
Thanks for taking a look, @poligilad-auto!

> 1. This is more related to the actual banner component, I noticed that the CTA is misaligned, it's too close to the right. Check this out. I don't know if this is something you can address.

This one is a tricky one, as it looks like this layout/behaviour has been a "feature" of the component for a long time, AFAICT. The specific offset here comes from [this line](https://github.com/Automattic/wp-calypso/blob/bf11d98185d96f623ae83d2332f57004195643cc/client/components/banner/style.scss#L218), which results in `margin-right: -6px;` being specified.

Given the ways in which the component is used, I think we'll need to handle that in a dedicated PR with some input from other teams.

> 2. What happens for users that have more than 1 eligible domain (I know this is an edge case 🙂)? Do we choose one of those domains to display the banner for?

We pick one -- we do give preference to the primary domain if that one is eligible, but we don't have a clear sequence beyond that. We could also sort by expiry date descending if we need to. (Though I checked our recent numbers for users who have more than one domain, and they're pretty small!)

daledupreez commented:
Side note: the unit test issue does look related.

semihakocer commented:
@daledupreez 

1. The placement of the banner looks great in the page architecture. The CTAs looked a bit weird  to me when they’re stacked like this. I think Poli also commented on a similar thing. I don’t know if we can override the general CTA dimensions, and padding on the left?

2. As we need a back-end sandbox I couldn't fully test it so a couple extra implementation questions: Are we going to auto-apply the 50% off for customers? If yes, we might want to mention it in the copy so they don’t try to search for a coupon. Are they going to see a coupon code and slashed out prices in checkout? (this ensures they complete the flow coming from here)

3. For copy, here are a couple proposals that might work with a stronger call to action
**1st line**
- Limited time sale! Get 50% off your first year with Google Workspace
- Get 50% off Google Workspace for a limited time! 
- Limited time only: 50% Off Google Workspace
**2nd line**
- Setup your custom mailbox @semihakocer.com and enable all the productivity tools Google Workspace offers. Discount will be auto-applied at checkout.
- Get Google Workspace and access all the benefits from setting up your custom email to
online storage, calendars, video meetings and more.
- Unlock your custom email @semihakocer.com and all the business tools Google offers while saving 50% off.
**CTA**
- Save 50%
- Get Your Mailbox
- Get Google Workspace

stephanethomas commented on client/my-sites/email/google-sale-banner/index.tsx:
```diff
@@ -0,0 +1,109 @@
+				'google-sale'
+			) }
+			title={ translate( 'Get %(discount)d% off Google Workspace for your domain!', {
+				args: { discount: googleWorkspaceProduct?.sale_coupon?.discount ?? 50 },
```
It may be better not to show this banner at all if that discount amount is not available.

daledupreez replied:
Done. I've added an explicit check for the `sale_coupon.discount` property.

daledupreez committed:
Address style issues/bugs

daledupreez committed:
Require a percentage discount

daledupreez committed:
Improve copy

daledupreez committed:
Fix unit test

daledupreez commented:
Thanks for the feedback, @semihakocer and @stephanethomas! Most of the items I'll cover here are specifically relating to your feedback, Semiha.

> The placement of the banner looks great in the page architecture. The CTAs looked a bit weird to me when they’re stacked like this. I think Poli also commented on a similar thing. I don’t know if we can override the general CTA dimensions, and padding on the left?

I've fixed this layout/alignment issue for the CTA. That was glaring once I _saw_ it!

> Are we going to auto-apply the 50% off for customers? If yes, we might want to mention it in the copy so they don’t try to search for a coupon. Are they going to see a coupon code and slashed out prices in checkout? (this ensures they complete the flow coming from here)

Yes, we are going to auto-apply the 50% off, and it will display correctly in checkout. There is no coupon needed or displayed in checkout, and the discounted price will also be displayed in the email comparison screen (which is where the CTA takes the user). This specific kind of sale coupon is always auto-applied, so the coupon code is not needed anywhere in the flow.

#### Copy

I've gone with the following copy, largely to keep things less cluttered/busy on mobile. I also don't think we need the checkout reference, because the discount will be visible before checkout as well.

**Get 50% off Google Workspace for a limited time!**
Set up your custom mailbox @example.com and enable all the productivity tools Google Workspace offers.

##### Mobile layout

<img width="379" alt="Screenshot 2021-11-03 at 21 17 26" src="https://user-images.githubusercontent.com/3376401/140181025-fef326b3-4bc0-40d0-9b44-e85b1284898f.png">

_Note: This screen capture was taken with a 40% off discount._

olaseni commented on client/my-sites/email/google-sale-banner/index.tsx:
```diff
@@ -0,0 +1,108 @@
+						domainName: domainForSale.name,
+					},
+					comment:
+						'%(domainName)s is a domain name, e.g. example.com; %(price)s is a formatted price, e.g. $3, £2.50',
```
We need to remove `%(price)s` from the translation comment since there's no `price` argument. Perhaps this was a vestige of an earlier refactor?

daledupreez replied:
Yep - I've removed that part of the comment in 1df979b.

olaseni commented on client/my-sites/email/google-sale-banner/index.tsx:
```diff
@@ -0,0 +1,108 @@
+				'google-sale'
+			) }
+			title={ translate( 'Get %(discount)d% off Google Workspace for a limited time!', {
+				args: { discount: googleWorkspaceProduct?.sale_coupon?.discount },
```
Do we still need the null-safe property access here?

daledupreez replied:
Removed in 1df979b.

saygunnyc commented:
As @poligilad-auto is AFK today, I wanted to share some thoughts, especially after seeing the mobile screenshot. 

I am aware some of this is related to the current state of things on the domains management page, and how content breaks on mobile. _– I honestly can't wait for the new domain management experience – pcYYhz-o2-p2 in production._ 

Regardless, for TODAY I have serious concerns about how someone will interact with this page and the offer because of the overall clutter and the competing and hard-to-read CTAs with too much copy over mobile.

Sharing a quick mockup, _–perhaps can also be seen as a provocation_ below for how we might potentially turn this around.

<img width="1029" alt="Screen Shot 2021-11-04 at 11 51 40" src="https://user-images.githubusercontent.com/78876323/140285065-d835cdbd-6044-43e2-ae34-1f090198e668.png">






Not sure if/how this would be feasible to apply at this point in the interim but still I think it would have a huge impact on both **the conversion of this offer** over this page and **the overall user experience of someone not interested in the offer but looking into managing their domains.**

olaseni reviewed (changes requested):
I have made a few refactor suggestions.

Also:

- The banner doesn't go away when I removed the sales coupon, even after a refresh. Hopefully this isn't specific to me.
- It looks like the CTA button in the tablet view needs a line-height rule. The text looks funny when it wraps.

<img width="126" alt="Screenshot 2021-11-04 at 10 10 54 AM" src="https://user-images.githubusercontent.com/277661/140287495-4fd95d53-3b14-4cd3-9a34-39773ce49b44.png">

- I know you mentioned it, but I think that Google Workspace on the Email Comparison page needs to be auto-expanded ( so as not prematurely loose users in that step ). This could be in another PR though.

daledupreez committed:
Updates based on code review:
 - Update line-height for button
 - Update stale/outdated translation comment
 - Remove unnecessary null-safe property access

daledupreez commented:
Thanks for the feedback, @saygunnyc and @olaseni!

> Regardless, for TODAY I have serious concerns about how someone will interact with this page and the offer because of the overall clutter and the competing and hard-to-read CTAs with too much copy over mobile.

Why not start simpler: can we just not show the sale banner on mobile? As I see it, all the banner is doing on this page is adding noise/clutter that prevents users from doing the things. they want to do.

We can also make the offer dismissible, but I feel that is also going to be imperfect without additional back-end support to remember the dismissal. So let's not get in our users' way.

> It looks like the CTA button in the tablet view needs a line-height rule. The text looks funny when it wraps.

I've increased the line-height for the button to give this room. Here is an updated tablet screenshot:

<img width="768" alt="Screenshot 2021-11-04 at 14 15 51" src="https://user-images.githubusercontent.com/3376401/140312909-021fe936-408c-4a96-bba6-396e9ee5edf3.png">

> The banner doesn't go away when I removed the sales coupon, even after a refresh. Hopefully this isn't specific to me.

AFAICT, this is due to _extremely_ aggressive caching for the coupons. I believe that adding a coupon busts the cache, but removing one doesn't, and I don't know how to do that manually. The general expectation is that coupons shouldn't be removed before their end date. Modifying the end date may actually be more effective, though I haven't tried that.

> I know you mentioned it, but I think that Google Workspace on the Email Comparison page needs to be auto-expanded ( so as not prematurely loose users in that step ). This could be in another PR though.

I think this should be handled in another PR.

daledupreez committed:
Hide banner on mobile

saygunnyc commented:
> Why not start simpler: can we not show the sale banner on mobile?

Ha! Honestly, I didn't think that could be a possibility. 😄 

The tablet view shared above is a lot better; perhaps we can simplify the CTA language to fit into a single line? It may not be necessary to repeat "Google Workspace".

daledupreez committed:
Add Tracks events and completely prevent render on mobile

daledupreez commented:
I've just updated the PR to change the "don't show on mobile" implementation, and to ensure that we log Tracks events for renders and clicks.

semihakocer commented:
>It may not be necessary to repeat "Google Workspace".
Alternatives to shorten the CTA:

Claim Now
Get Today
Use Discount

semihakocer commented:
Other changes looks good to me @daledupreez

olaseni commented on client/my-sites/email/google-sale-banner/index.tsx:
```diff
@@ -56,6 +58,10 @@ const GoogleSaleBanner: FunctionComponent< GoogleSaleBannerProps > = ( { domains
 	);
 
+	if ( isMobile ) {
+		return null;
```
Curious: Why did you switch to a no-style solution?

daledupreez replied:
I wanted to be sure that we wouldn't trigger the Tracks events -- I suspected that the `TrackComponentView` logic in the `Banner` component would still trigger an event even if the CSS was set to `display: none;`. So I want to be extra-sure we were tracking things correctly.

olaseni commented:
> Modifying the end date may actually be more effective, though I haven't tried that.

I can confirm that modifying this works. When the coupon is past its end/expiration date, it doesn't display. 

Hopefully we can remember this in the future? Or it may not even arise if the campaign expires naturally.

olaseni reviewed (approved):
There's some feedback from @semihakocer but I don't think that's blocking. Looks good.

I left some comments about the sticky banner issue.

saygunnyc commented:
> Claim Now

Sounds good to me!

daledupreez committed:
Change CTA to 'Claim Now'

daledupreez commented:
I've just updated the CTA to "Claim Now".

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/6946503

Thank you @daledupreez for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/6946503).
